Deprecate `RustWriter.compileAndTest`

david-perez opened the pull request:
To dissuade contributors from introducing more usages; we want to remove
this extension function. See #3193.

----

_By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice._

david-perez commented:
As per https://github.com/smithy-lang/smithy-rs/issues/3193#issuecomment-1812509755, we should only deprecate `RustWriter.compileAndTest`. There's some value in `TestWriterDelegator.compileAndTest`.

drganjoo commented on codegen-core/src/main/kotlin/software/amazon/smithy/rust/codegen/core/testutil/Rust.kt:
```diff
@@ -320,6 +320,7 @@ fun FileManifest.printGeneratedFiles() {
  * should generally be set to `false` to avoid invalidating the Cargo cache between
  * every unit test run.
  */
+@Deprecated("use `{client,server}IntegrationTest instead, which is simpler to use, or `TestWriterDelegator.compileAndTest` if your generated code is simple and focused and don't require to codegen a full SDK: https://github.com/smithy-lang/smithy-rs/issues/3193")
```
The ending tick is missing:

"use `{client,server}IntegrationTest` <---

drganjoo commented on codegen-core/src/main/kotlin/software/amazon/smithy/rust/codegen/core/testutil/Rust.kt:
```diff
@@ -387,6 +388,7 @@ fun String.shouldParseAsRust() {
 /**
  * Compiles the contents of the given writer (including dependencies) and runs the tests
  */
+@Deprecated("use `{client,server}IntegrationTest instead, which is simpler to use, or `TestWriterDelegator.compileAndTest` if your generated code is simple and focused and don't require to codegen a full SDK: https://github.com/smithy-lang/smithy-rs/issues/3193")
```
Ending tick missing.

david-perez commented:
@drganjoo Thanks! I've added the backtick, and removed the deprecation message from `TestWriterDelegator.compileAndTest`, as I'd said I do in https://github.com/smithy-lang/smithy-rs/pull/3195#issuecomment-1829690546. I've also freshened up the branch. This is now ready to be merged.

david-perez commented:
PR title and description have also been updated.

jdisanti commented:
Going to close this since there are valid use cases for using compileAndTest for focused code generating classes.
